crates/kubelet/src/server.rs

```diff
@@ -1,189 +1,110 @@
-                    "follow" => (),
-                    s => warn!("Unknown query parameter: {}={}", s, value),
-                }
-            }
-            get_container_logs(
-                &*provider,
-                &req,
-                (*namespace).to_string(),
-                (*pod).to_string(),
-                (*container).to_string(),
-                tail,
-                follow,
-            )
-            .await
-        }
-        (&Method::POST, [_, "exec", _, _, _]) => post_exec(&*provider, &req),
-        _ => Response::builder()
-            .status(StatusCode::NOT_FOUND)
-            .body(Body::from("Not Found"))
-            .unwrap(),
-    };
-
-    Ok(response)
-}
-
-/// Return a simple status message
-fn get_ping() -> Response<Body> {
-    Response::new(Body::from("this is the Krustlet HTTP server"))
-}
-
-/// Get the logs from the running WASM module
+/// Get the logs from the running container.
 ///
 /// Implements the kubelet path /containerLogs/{namespace}/{pod}/{container}
-async fn get_container_logs<T: Provider + Sync>(
-    provider: &T,
-    req: &Request<Body>,
+async fn get_container_logs<T: 'static + Provider + Send + Sync>(
+    provider: Arc<T>,
     namespace: String,
     pod: String,
     container: String,
-    tail: Option<usize>,
-    follow: bool,
-) -> Response<Body> {
+    opts: LogOptions,
+) -> Result<Response<Body>, Infallible> {
     debug!(
-        "Got container log request for container {} in pod {} in namespace {}. tail: {:?}, follow: {}",
-        container, pod, namespace, tail, follow
+        "Got container log request for container {} in pod {} in namespace {}. Options: {:?}.",
+        container, pod, namespace, opts
     );
-    if namespace.is_empty() || pod.is_empty() || container.is_empty() {
-        return Response::builder()
-            .status(StatusCode::NOT_FOUND)
-            .body(Body::from(format!(
-                "Resource {} not found",
-                req.uri().path()
-            )))
-            .unwrap();
-    }
-    let (sender, log_body) = hyper::Body::channel();
-    let log_sender = LogSender::new(sender, tail, follow);
+    let (sender, log_body) = Body::channel();
+    let log_sender = LogSender::new(sender, opts);
 
     match provider.logs(namespace, pod, container, log_sender).await {
-        Ok(()) => Response::new(log_body),
+        Ok(()) => Ok(Response::new(log_body)),
         Err(e) => {
             error!("Error fetching logs: {}", e);
-            let mut res = Response::new(Body::from(format!("Server error: {}", e)));
-            *res.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
             if e.is::<NotImplementedError>() {
-                res = Response::new(Body::from("Not Implemented"));
-                *res.status_mut() = StatusCode::NOT_IMPLEMENTED;
+                return_with_code(
+                    StatusCode::NOT_IMPLEMENTED,
+                    format!("Logs not implemented in provider."),
```

Comment:
No need to use the `format!` macro here. `String::from`, `.to_owned()` or `.to_string()`  would be preferable